### PR TITLE
feat(core): add payments post_session_tokens flow

### PR DESCRIPTION
**Pull request by swangi-kumari**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [x] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
1. Add support for a new endpoint post_session_tokens which can be called before the confirm call.

Use Case
- In PayPal, after the session call, when we click the PayPal button, we need to create an order. Previously, this was done in the `/confirm` call, and in response, we received the `order_id` / `connector_transaction_id`. We would then send this information to the HS SDK, which forwarded it to PayPal. PayPal used this data to render their SDK, allowing the user to log in and modify the shipping address. The user would then click "Complete Purchase," triggering the OnApprove callback, which was handled by the /complete_authorize call.
- The /confirm call happens at the point of order creation, but we receive the shipping address after that. At that stage, we make a call to TaxJar, and the order's tax amount is added. However, since the /confirm call has already happened, we can't change/update the amount at that point. This was the reason for creating the new flow, where the order creation happens in the /post_session_tokens call, and the /confirm call occurs after tax calculation.
- After this refactoring, the CreateOrder step will be handled in the /post_session_tokens call, and the OnApprove action will be handled in the /confirm call.

2. Populate Dynamic Fields for Paypal

### Additional Changes

- [x] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?
1. Create Merchant Account
```
curl --location 'http://localhost:8080/accounts' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: test_admin' \
--data-raw '{
  "merchant_id": "merchant_1728546485",
  "locker_id": "m0010",
  "merchant_name": "NewAge Retailer",
  "merchant_details": {
    "primary_contact_person": "John Test",
    "primary_email": "JohnTest@test.com",
    "primary_phone": "sunt laborum",
    "secondary_contact_person": "John Test2",
    "secondary_email": "JohnTest2@test.com",
    "secondary_phone": "cillum do dolor id",
    "website": "https://www.example.com",
    "about_business": "Online Retail with a wide selection of organic products for North America",
    "address": {
      "line1": "1467",
      "line2": "Harrison Street",
      "line3": "Harrison Street",
      "city": "San Fransico",
      "state": "California",
      "zip": "94122",
      "country": "US"
    }
  },
  "return_url": "https://google.com/success",
  "webhook_details": {
    "webhook_version": "1.0.1",
    "webhook_username": "ekart_retail",
    "webhook_password": "password_ekart@123",
    "payment_created_enabled": true,
    "payment_succeeded_enabled": true,
    "payment_failed_enabled": true
  },
  "sub_merchants_enabled": false,
  "metadata": {
    "city": "NY",
    "unit": "245"
  },
  "primary_business_details": [
    {
      "country": "US",
      "business": "food"
    }
  ]
}'
```
2. Create API Key
```
curl --location 'http://localhost:8080/api_keys/merchant_1728540504' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: test_admin' \
--data '{
  "name": "API Key 1",
  "description": null,
  "expiration": "2025-09-23T01:02:03.000Z"
}'
```
3. Create MCA
```
curl --location 'http://localhost:8080/account/merchant_1728540504/connectors?=' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: test_admin' \
--data '{
    "connector_type": "payment_processor",
    "connector_name": "paypal",
    "connector_account_details": {
        "auth_type": "BodyKey",
        "api_key": "_",
        "key1": "_"
    },
    "test_mode": false,
    "disabled": false,
    "payment_methods_enabled": [
        {
            "payment_method": "card",
            "payment_method_types": [
                {
                    "payment_method_type": "credit",
                    "card_networks": [
                        "Visa",
                        "Mastercard"
                    ],
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": true
                },
                {
                    "payment_method_type": "debit",
                    "card_networks": [
                        "Visa",
                        "Mastercard"
                    ],
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": true
                }
            ]
        },
        {
            "payment_method": "wallet",
            "payment_method_types": [
                {
                    "payment_method_type": "paypal",
                    "payment_experience": "redirect_to_url",
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": false,
                    "installment_payment_enabled": false
                },
                {
                    "payment_method_type": "paypal",
                    "payment_experience": "invoke_sdk_client",
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": false,
                    "installment_payment_enabled": false
                }
                
            ]
        },
        {
            "payment_method": "bank_redirect",
            "payment_method_types": [
                {
                    "payment_method_type": "giropay",
                    "recurring_enabled": true,
                    "installment_payment_enabled": true,
                    "payment_experience": "redirect_to_url"
                },
                {
                    "payment_method_type": "ideal",
                    "recurring_enabled": true,
                    "installment_payment_enabled": true,
                    "payment_experience": "redirect_to_url"
                },
                {
                    "payment_method_type": "eps",
                    "recurring_enabled": true,
                    "installment_payment_enabled": true,
                    "payment_experience": "redirect_to_url"
                },
                {
                    "payment_method_type": "sofort",
                    "recurring_enabled": true,
                    "installment_payment_enabled": true,
                    "payment_experience": "redirect_to_url"
                }
            ]
        }
    ],
     "metadata": {
        "paypal_sdk": {
            "client_id": "___"
        }
    },
    "business_country": "US",
    "business_label": "food"
}'
```
4. Create Payment 
```
curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_MTxKOCiP8GOiXv955xYCMuW0MUzl1anEbVVkxBBR87wfDUObqYqOsCqpogsFFnCd' \
--data-raw '{
    "amount": 6540,
    "currency": "USD",
    "confirm": false,
    "capture_method": "automatic",
    "capture_on": "2022-09-10T10:11:12Z",
    "amount_to_capture": 6540,
    "customer_id": "StripeCustomer",
    "email": "guest@example.com",
    "name": "John Doe",
    "phone": "999999999",
    "phone_country_code": "+1",
    "description": "Its my first payment request",
    "authentication_type": "no_three_ds",
    "return_url": "https://duck.com",
    "billing": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "PiX"
        }
    },
    "shipping": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "PiX"
        }
    },
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "metadata": {
        "udf1": "value1",
        "new_customer": "true",
        "login_date": "2019-09-10T10:11:12Z"
    }
}'
```
5.  Create Session token
```
curl --location 'http://localhost:8080/payments/session_tokens' \
--header 'Content-Type: application/json' \
--header 'api-key: pk_dev_10cfee22267a4184a3d45665920788ea' \
--data '{
    "payment_id": "pay_J3IXGvol8nNvUkSRTcwQ",
    "wallets": [],
    "client_secret": "pay_J3IXGvol8nNvUkSRTcwQ_secret_VN76rK4oCqTghQ1kwNcO"
}'
```
Response
```
{
    "payment_id": "pay_mcVzUMZws5KmEKbGkZ0g",
    "client_secret": "pay_mcVzUMZws5KmEKbGkZ0g_secret_RMWlmy1jSdlp9VOMX1qc",
    "session_token": [
        {
            "wallet_name": "paypal",
            "connector": "paypal",
            "session_token": "ASKAGh2WXgqfQ5TzjpZzLsfhVGlFbjq5VrV5IOX8KXDD2N_XqkGeYNDkWyr_UXnfhXpEkABdmP284b_2",
            "sdk_next_action": {
                "next_action": "post_session_tokens"
            }
        }
    ]
}
```
In Response Check for Next action - It should be `post_session_tokens`

6. List Payment Methods for merchant - Check for required fields (shipping_details)
```
curl --location 'http://localhost:8080/account/payment_methods?client_secret=pay_J3IXGvol8nNvUkSRTcwQ_secret_VN76rK4oCqTghQ1kwNcO' \
--header 'Accept: application/json' \
--header 'api-key: pk_dev_10cfee22267a4184a3d45665920788ea'
```
Response
```
{
    "redirect_url": "https://google.com/success",
    "currency": "USD",
    "payment_methods": [
        {
            "payment_method": "wallet",
            "payment_method_types": [
                {
                    "payment_method_type": "paypal",
                    "payment_experience": [
                        {
                            "payment_experience_type": "invoke_sdk_client",
                            "eligible_connectors": [
                                "paypal"
                            ]
                        },
                        {
                            "payment_experience_type": "redirect_to_url",
                            "eligible_connectors": [
                                "paypal"
                            ]
                        }
                    ],
                    "card_networks": null,
                    "bank_names": null,
                    "bank_debits": null,
                    "bank_transfers": null,
                    "required_fields": {
                        "shipping.address.country": {
                            "required_field": "shipping.address.country",
                            "display_name": "country",
                            "field_type": {
                                "user_shipping_address_country": {
                                    "options": [
                                        "ALL"
                                    ]
                                }
                            },
                            "value": "US"
                        },
                        "shipping.address.first_name": {
                            "required_field": "shipping.address.first_name",
                            "display_name": "shipping_first_name",
                            "field_type": "user_shipping_name",
                            "value": "PiX"
                        },
                        "shipping.address.zip": {
                            "required_field": "shipping.address.zip",
                            "display_name": "zip",
                            "field_type": "user_shipping_address_pincode",
                            "value": "94122"
                        },
                        "shipping.address.state": {
                            "required_field": "shipping.address.state",
                            "display_name": "state",
                            "field_type": "user_shipping_address_state",
                            "value": "California"
                        },
                        "shipping.address.city": {
                            "required_field": "shipping.address.city",
                            "display_name": "city",
                            "field_type": "user_shipping_address_city",
                            "value": "San Fransico"
                        },
                        "shipping.address.last_name": {
                            "required_field": "shipping.address.last_name",
                            "display_name": "shipping_last_name",
                            "field_type": "user_shipping_name",
                            "value": null
                        },
                        "shipping.address.line1": {
                            "required_field": "shipping.address.line1",
                            "display_name": "line1",
                            "field_type": "user_shipping_address_line1",
                            "value": "1467"
                        }
                    },
                    "surcharge_details": null,
                    "pm_auth_connector": null
                }
            ]
        },
            ],
    "mandate_payment": null,
    "merchant_name": "NewAge Retailer",
    "show_surcharge_breakup_screen": false,
    "payment_type": "normal",
    "request_external_three_ds_authentication": false,
    "collect_shipping_details_from_wallets": false,
    "collect_billing_details_from_wallets": false,
    "is_tax_calculation_enabled": false
}
```
7. Create Post Session Tokens for Paypal
```
curl --location 'http://localhost:8080/payments/pay_J3IXGvol8nNvUkSRTcwQ/post_session_tokens' \
--header 'Content-Type: application/json' \
--header 'api-key: pk_dev_10cfee22267a4184a3d45665920788ea' \
--data '{
    "client_secret": "pay_J3IXGvol8nNvUkSRTcwQ_secret_VN76rK4oCqTghQ1kwNcO",
    "payment_method_type": "paypal",
    "payment_method": "wallet"
}'
```
Response
```
 {
    "payment_id": "pay_mcVzUMZws5KmEKbGkZ0g",
    "next_action": {
        "type": "invoke_sdk_client",
        "next_action_data": {
            "next_action": "confirm",
            "order_id": "2KW93090954330301"
        }
    }
}
```
In Response Check for Next action - It should be confirm

8. Confirm Request 
```
curl 'http://localhost:8080/payments/pay_6oSZ2utmkjgdOlfrZ53E/confirm' \
-X 'POST' \
-H 'Content-Type: application/json' \
-H 'Accept-Encoding: gzip, deflate' \
-H 'Sec-Fetch-Mode: cors' \
-H 'x-payment-confirm-source: sdk' \
-H 'x-client-source: ExpressCheckoutElement' \
-H 'api-key: pk_dev_43922dfcb2694e09929891b0f4f7b6dc' \
-H 'x-browser-version: 605.1.15' \
--data-binary '{"client_secret":"pay_6oSZ2utmkjgdOlfrZ53E_secret_EFrF7Izc98NHygNyG51x","return_url":"http://localhost:9060","retry_action":"manual_retry","shipping":{"address":{"country":"US","state":"California","zip":"94588","first_name":"Jagan","city":"Pleasanton","last_name":"Elavarasan","line1":"4141"}},"payment_method":"wallet","payment_method_type":"paypal","payment_experience":"invoke_sdk_client","connector":["paypal"],"payment_method_data":{"wallet":{"paypal_sdk":{"token":""}}},"browser_info":{"user_agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.1.2 Safari/605.1.15","accept_header":"text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8","language":"en-US","color_depth":24,"screen_height":1117,"screen_width":1728,"time_zone":-330,"java_enabled":true,"java_script_enabled":true},"payment_type":"normal"}'
```
Response
```{
    "payment_id": "pay_6oSZ2utmkjgdOlfrZ53E",
    "merchant_id": "merchant_1728844379",
    "status": "succeeded",
    "amount": 2999,
    "net_amount": 2999,
    "amount_capturable": 0,
    "amount_received": 2999,
    "connector": "paypal",
    "client_secret": "pay_6oSZ2utmkjgdOlfrZ53E_secret_EFrF7Izc98NHygNyG51x",
    "created": "2024-10-14T09:09:36.340Z",
    "currency": "USD",
    "customer_id": "hyperswitch_sdk_demo_id",
    "customer": {
        "id": "hyperswitch_sdk_demo_id",
        "name": null,
        "email": "hyperswitch_sdk_demo_id@gmail.com",
        "phone": null,
        "phone_country_code": null
    },
    "description": "Hello this is description",
    "refunds": null,
    "disputes": null,
    "mandate_id": null,
    "mandate_data": null,
    "setup_future_usage": null,
    "off_session": null,
    "capture_on": null,
    "capture_method": "automatic",
    "payment_method": "wallet",
    "payment_method_data": {
        "wallet": {},
        "billing": null
    },
    "payment_token": null,
    "shipping": {
        "address": {
            "city": "Pleasanton",
            "country": "US",
            "line1": "4141",
            "line2": "Amsterdam Ave",
            "line3": "alsksoe",
            "zip": "94588",
            "state": "California",
            "first_name": "Jagan",
            "last_name": "Elavarasan"
        },
        "phone": {
            "number": "5551234",
            "country_code": "+1"
        },
        "email": "customer@email.us"
    },
    "billing": {
        "address": {
            "city": "San Fransico",
            "country": "US",
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "zip": "94122",
            "state": "California",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "_",
            "country_code": "+91"
        },
        "email": "swangi@getMaxListeners.com"
    },
    "order_details": [
        {
            "brand": null,
            "amount": 2999,
            "category": null,
            "quantity": 1,
            "product_id": null,
            "product_name": "Apple iPhone 15",
            "product_type": null,
            "sub_category": null,
            "product_img_link": null,
            "product_tax_code": null,
            "requires_shipping": null
        }
    ],
    "email": "hyperswitch_sdk_demo_id@gmail.com",
    "name": null,
    "phone": null,
    "return_url": "http://localhost:9060/",
    "authentication_type": "three_ds",
    "statement_descriptor_name": null,
    "statement_descriptor_suffix": null,
    "next_action": null,
    "cancellation_reason": null,
    "error_code": null,
    "error_message": null,
    "unified_code": null,
    "unified_message": null,
    "payment_experience": "invoke_sdk_client",
    "payment_method_type": "paypal",
    "connector_label": null,
    "business_country": null,
    "business_label": "default",
    "business_sub_label": null,
    "allowed_payment_method_types": null,
    "ephemeral_key": null,
    "manual_retry_allowed": false,
    "connector_transaction_id": "1SV0561343891340C",
    "frm_message": null,
    "metadata": {
        "udf1": "value1",
        "login_date": "2019-09-10T10:11:12Z",
        "new_customer": "true"
    },
    "connector_metadata": null,
    "feature_metadata": null,
    "reference_id": "pay_6oSZ2utmkjgdOlfrZ53E_1",
    "payment_link": null,
    "profile_id": "pro_03iBylVoOpuldQsVZKtO",
    "surcharge_details": null,
    "attempt_count": 1,
    "merchant_decision": null,
    "merchant_connector_id": "mca_frKbF7wxB3Rt7tAFbpJA",
    "incremental_authorization_allowed": null,
    "authorization_count": null,
    "incremental_authorizations": null,
    "external_authentication_details": null,
    "external_3ds_authentication_attempted": false,
    "expires_on": "2024-10-14T09:24:36.340Z",
    "fingerprint": null,
    "browser_info": {
        "language": "en-US",
        "time_zone": -330,
        "ip_address": "::1",
        "user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.1.2 Safari/605.1.15",
        "color_depth": 24,
        "java_enabled": true,
        "screen_width": 1728,
        "accept_header": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8",
        "screen_height": 1117,
        "java_script_enabled": true
    },
    "payment_method_id": null,
    "payment_method_status": null,
    "updated": "2024-10-14T09:10:19.289Z",
    "charges": null,
    "frm_metadata": null,
    "merchant_order_reference_id": null,
    "order_tax_amount": null,
    "connector_mandate_id": null
}
```

For end to end testing, we need to do the payment via SDK

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/router/src/core/payments/operations/payment_response.rs [hunk 1]
```diff
@@ -687,6 +687,84 @@ impl<F: Clone> PostUpdateTracker<F, PaymentData<F>, types::SdkPaymentsSessionUpd
...
+            }) => {
+                let connector_transaction_id = match resource_id {
+                    types::ResponseId::NoResponseId => None,
+                    types::ResponseId::ConnectorTransactionId(id)
+                    | types::ResponseId::EncodedData(id) => Some(id),
+                };
+                let m_db = db.clone().store;
+                let payment_attempt_update =
+                    storage::PaymentAttemptUpdate::PostSessionTokensUpdate {
+                        connector_transaction_id: connector_transaction_id.clone(),
```
Comment (Narayanbhat166) [inline crates/router/src/core/payments/operations/payment_response.rs:731]: we should not update `connector_transaction_id` in this flow. There can be cases where confirm happens through a different payment method data and a new connector might be selected, we will have the wrong mapping of connector and transaction id.

File: crates/router/src/core/payments/operations/payment_response.rs [hunk 2]
```diff
@@ -687,6 +687,84 @@ impl<F: Clone> PostUpdateTracker<F, PaymentData<F>, types::SdkPaymentsSessionUpd
...
+                payment_data.payment_attempt = updated_payment_attempt;
+            }
+            Err(err) => {
+                Err(errors::ApiErrorResponse::ExternalConnectorError {
+                    code: err.code,
+                    message: err.message,
+                    connector: payment_data
+                        .payment_attempt
+                        .connector
+                        .clone()
```
Comment (Narayanbhat166) [inline crates/router/src/core/payments/operations/payment_response.rs:752]: we will not store the connector in this case. We should not update the error message in this case. This should be a bad request error.

**Review by Narayanbhat166 — CHANGES_REQUESTED** [review]

File: crates/diesel_models/src/payment_attempt.rs [hunk 3]
```diff
@@ -475,6 +475,11 @@ pub enum PaymentAttemptUpdate {
         unified_message: Option<String>,
         connector_transaction_id: Option<String>,
     },
+    PostSessionTokensUpdate {
+        updated_by: String,
+        connector_transaction_id: Option<String>,
```
Comment (Narayanbhat166) [inline crates/diesel_models/src/payment_attempt.rs:480]: do not update this

File: crates/router/src/types/api/payments.rs [hunk 4]
```diff
@@ -9,7 +9,8 @@ pub use api_models::payments::{
     PaymentsApproveRequest, PaymentsCancelRequest, PaymentsCaptureRequest,
     PaymentsCompleteAuthorizeRequest, PaymentsDynamicTaxCalculationRequest,
     PaymentsDynamicTaxCalculationResponse, PaymentsExternalAuthenticationRequest,
-    PaymentsIncrementalAuthorizationRequest, PaymentsManualUpdateRequest, PaymentsRedirectRequest,
+    PaymentsIncrementalAuthorizationRequest, PaymentsManualUpdateRequest,
```
Comment (Narayanbhat166) [inline crates/router/src/types/api/payments.rs:12]: @SanchithHegde what do you think about such re exports? This adds friction when a developer is trying to move his way to from where and how this struct has been imported. In my opinion we should always qualify specific imports from the parent module. This might take some time, but I think it will be better to move away from this.

File: crates/router/src/core/payments/operations/payment_post_session_tokens.rs [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+            state,
+            payment_intent.billing_address_id.clone(),
+            key_store,
+            &payment_intent.payment_id,
+            merchant_id,
+            merchant_account.storage_scheme,
+        )
+        .await?;
+
+        let payment_method_billing = helpers::get_address_by_id(
```
[the remark this answers is not in the document]
Comment (sai-harsha-vardhan) [reply crates/router/src/core/payments/operations/payment_post_session_tokens.rs]: Billing address would be there if merchant has sent it in payments create request right? @Narayanbhat166

File: crates/router/src/core/payments/operations/payment_post_session_tokens.rs [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+            .change_context(errors::ApiErrorResponse::InternalServerError)
+            .attach_printable("'profile_id' not set in payment intent")?;
+
+        let business_profile = db
+            .find_business_profile_by_profile_id(key_manager_state, key_store, profile_id)
+            .await
+            .to_not_found_response(errors::ApiErrorResponse::ProfileNotFound {
+                id: profile_id.get_string_repr().to_owned(),
+            })?;
+        payment_attempt.payment_method = Some(request.payment_method);
```
Comment (sai-harsha-vardhan) [inline crates/router/src/core/payments/operations/payment_post_session_tokens.rs:101]: Is this change required?

File: crates/router/src/core/payments/operations/payment_response.rs [hunk 7]
```diff
@@ -687,6 +687,67 @@ impl<F: Clone> PostUpdateTracker<F, PaymentData<F>, types::SdkPaymentsSessionUpd
...
+        mut payment_data: PaymentData<F>,
+        router_data: types::RouterData<
+            F,
+            types::PaymentsPostSessionTokensData,
+            types::PaymentsResponseData,
+        >,
+        _key_store: &domain::MerchantKeyStore,
+        storage_scheme: enums::MerchantStorageScheme,
+        _locale: &Option<String>,
+        #[cfg(all(feature = "v1", feature = "dynamic_routing"))] _routable_connector: Vec<
```
Comment (sai-harsha-vardhan) [inline crates/router/src/core/payments/operations/payment_response.rs:708]: These feature flags are redundant as we have v1 on top level

File: crates/router/src/core/payments/transformers.rs [hunk 8]
```diff
@@ -516,6 +516,41 @@ where
...
+        let papal_sdk_next_action =
+            paypal_sdk_next_steps_check(payment_data.get_payment_attempt().clone())?;
+        let next_action = papal_sdk_next_action.map(|paypal_next_action_data| {
+            api_models::payments::NextActionData::InvokeSdkClient {
+                next_action_data: paypal_next_action_data,
+            }
+        });
+        Ok(services::ApplicationResponse::JsonWithHeaders((
+            Self {
+                payment_id: payment_data.get_payment_attempt().payment_id.clone(),
```
Comment (sai-harsha-vardhan) [inline crates/router/src/core/payments/transformers.rs:546]: You can get payment_id from payment_intent here

File: crates/router_env/src/logger/types.rs [hunk 9]
```diff
@@ -498,6 +498,8 @@ pub enum Flow {
     PaymentsManualUpdate,
     /// Dynamic Tax Calcultion
     SessionUpdateTaxCalculation,
+    /// Payments create order flow
```
Comment (sai-harsha-vardhan) [inline crates/router_env/src/logger/types.rs:501]:
```suggestion
    /// Payments post session tokens flow
```

File: crates/router/src/core/payments/flows/session_flow.rs [hunk 10]
```diff
@@ -955,7 +955,7 @@ fn create_paypal_sdk_session_token(
                     connector: connector.connector_name.to_string(),
                     session_token: paypal_sdk_data.data.client_id,
                     sdk_next_action: payment_types::SdkNextAction {
-                        next_action: payment_types::NextActionCall::Confirm,
+                        next_action: payment_types::NextActionCall::PostSessionTokens,
```
[the remark this answers is not in the document]
Comment (swangi-kumari) [reply crates/router/src/core/payments/flows/session_flow.rs:958]: Yes, this will go in sandbox after SDK sandbox deployment.

File: crates/router/src/connector/paypal.rs [hunk 11]
```diff
@@ -692,8 +800,20 @@ impl ConnectorIntegration<api::Authorize, types::PaymentsAuthorizeData, types::P
         req: &types::PaymentsAuthorizeRouterData,
         connectors: &settings::Connectors,
     ) -> CustomResult<Option<services::Request>, errors::ConnectorError> {
-        Ok(Some(
-            services::RequestBuilder::new()
+        let payment_method_data = req.request.payment_method_data.clone();
+        let req = match payment_method_data {
+            domain::PaymentMethodData::Wallet(domain::WalletData::PaypalSdk(_)) => {
+                services::RequestBuilder::new()
+                    .method(services::Method::Post)
```
[the remark this answers is not in the document]
Comment (swangi-kumari) [reply crates/router/src/connector/paypal.rs:804]: We are using this flow to authorize the payment.
It should be Post Method - [doc](https://developer.paypal.com/docs/api/orders/v2/#orders_authorize)

File: crates/router/src/connector/paypal.rs [hunk 12]
```diff
@@ -649,6 +652,94 @@ impl
...
+        Ok(RequestContent::Json(Box::new(connector_req)))
+    }
+
+    fn handle_response(
+        &self,
+        data: &types::PaymentsPostSessionTokensRouterData,
+        event_builder: Option<&mut ConnectorEvent>,
+        res: Response,
+    ) -> CustomResult<types::PaymentsPostSessionTokensRouterData, errors::ConnectorError> {
+        let response: paypal::PaypalRedirectResponse = res
```
[the remark this answers is not in the document]
Comment (swangi-kumari) [reply crates/router/src/connector/paypal.rs:721]: The response which get for PaypaSdk Create Order , matches the struct PaypalRedirectResponse.

File: crates/router/src/core/payments/transformers.rs [hunk 13]
```diff
@@ -2078,6 +2114,45 @@ impl<F: Clone> TryFrom<PaymentAdditionalData<'_, F>> for types::SdkPaymentsSessi
...
+            .surcharge_details
+            .as_ref()
+            .map(|surcharge_details| surcharge_details.get_total_surcharge_amount())
+            .unwrap_or_default();
+        let shipping_cost = payment_data
+            .payment_intent
+            .shipping_cost
+            .unwrap_or_default();
+        // amount here would include amount, surcharge_amount and shipping_cost
+        let amount = payment_data.payment_intent.amount + shipping_cost + surcharge_amount;
```
[the remark this answers is not in the document]
Comment (Narayanbhat166) [reply crates/router/src/core/payments/transformers.rs:2142]: @swangi-kumari we should include order_tax amount here right?

**Review by Narayanbhat166 — APPROVED** [review]

**Review by deepanshu-iiitu — APPROVED** [review]

**Review by tsdk02 — APPROVED** [review]

File: crates/api_models/src/payments.rs [hunk 14]
```diff
@@ -4873,6 +4874,34 @@ pub struct PaymentsSessionRequest {
     pub merchant_connector_details: Option<admin::MerchantConnectorDetailsWrap>,
 }
 
+#[derive(Debug, serde::Deserialize, serde::Serialize, Clone, ToSchema)]
+pub struct PaymentsPostSessionTokensRequest {
+    /// The unique identifier for the payment
+    #[serde(skip_deserializing)]
+    #[schema(value_type = String)]
+    pub payment_id: id_type::PaymentId,
```
Comment (SanchithHegde) [inline crates/api_models/src/payments.rs:4882]: If we're not obtaining this from the request (not deserializing it), why are we including it in the request?

File: crates/router/src/services/authentication.rs [hunk 15]
```diff
@@ -1803,6 +1803,12 @@ impl ClientSecretFetch for PaymentMethodListRequest {
     }
 }
 
+impl ClientSecretFetch for payments::PaymentsPostSessionTokensRequest {
+    fn get_client_secret(&self) -> Option<&String> {
+        Some(self.client_secret.peek())
```
Comment (apoorvdixit88) [inline crates/router/src/services/authentication.rs:1808]: may be 
```suggestion
       self.client_secret.as_ref()
```

**Review by apoorvdixit88 — APPROVED** [review]

File: crates/router/src/connector/paypal.rs [hunk 16]
```diff
@@ -649,6 +652,94 @@ impl
...
+        req: &types::PaymentsPostSessionTokensRouterData,
+        connectors: &settings::Connectors,
+    ) -> CustomResult<Option<services::Request>, errors::ConnectorError> {
+        Ok(Some(
+            services::RequestBuilder::new()
+                .method(services::Method::Post)
+                .url(&types::PaymentsPostSessionTokensType::get_url(
+                    self, req, connectors,
+                )?)
+                .headers(types::PaymentsPostSessionTokensType::get_headers(
```
Comment (SanchithHegde) [inline crates/router/src/connector/paypal.rs:690]: Can we call `attach_default_headers()` as well?

File: crates/router/src/connector/paypal.rs [hunk 17]
```diff
@@ -692,8 +799,20 @@ impl ConnectorIntegration<api::Authorize, types::PaymentsAuthorizeData, types::P
...
-            services::RequestBuilder::new()
+        let payment_method_data = req.request.payment_method_data.clone();
+        let req = match payment_method_data {
+            domain::PaymentMethodData::Wallet(domain::WalletData::PaypalSdk(_)) => {
+                services::RequestBuilder::new()
+                    .method(services::Method::Post)
+                    .url(&types::PaymentsAuthorizeType::get_url(
+                        self, req, connectors,
+                    )?)
+                    .headers(types::PaymentsAuthorizeType::get_headers(
```
Comment (SanchithHegde) [inline crates/router/src/connector/paypal.rs:810]: Same here.

File: crates/router/src/connector/paypal/transformers.rs [hunk 18]
```diff
@@ -136,6 +151,22 @@ impl From<&PaypalRouterData<&types::PaymentsAuthorizeRouterData>> for ItemDetail
...
 }
 
+impl From<&PaypalRouterData<&types::PaymentsPostSessionTokensRouterData>> for ItemDetails {
+    fn from(item: &PaypalRouterData<&types::PaymentsPostSessionTokensRouterData>) -> Self {
+        Self {
+            name: format!(
+                "Payment for invoice {}",
+                item.router_data.connector_request_reference_id
+            ),
+            quantity: 1,
```
Comment (SanchithHegde) [inline crates/router/src/connector/paypal/transformers.rs:161]: Any reason this is hardcoded?

File: crates/router/src/core/payments/flows/session_flow.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (SanchithHegde) [reply crates/router/src/core/payments/flows/session_flow.rs:958]: > Yes, this will go in sandbox after SDK sandbox deployment.

What's our strategy for production rollout of this change, would it be rolled out after SDK deployment on production as well?

File: crates/router/src/types/api/payments.rs (hunk 4 above)
Comment (SanchithHegde) [reply crates/router/src/types/api/payments.rs:12]: I believe this was kept this way when we moved out API models to a separate crate, to reduce the number of changes elsewhere in code. But yeah, I think we should avoid such public re-exports.

Of course, this can be taken up in a separate PR.

File: crates/api_models/src/payments.rs (hunk 14 above)
Comment (swangi-kumari) [reply crates/api_models/src/payments.rs:4882]: This is for ApiEventMetrics and ApiLocking

Comment (SanchithHegde) [reply crates/api_models/src/payments.rs:4882]: @Narayanbhat166 Should we consider having an "internal" struct for such purposes, or are we okay with this practice?

**Comment by swangi-kumari** [comment]
> What's our strategy for production rollout of this change, would it be rolled out after SDK deployment on production as well?
Yes we will have SDK production deployment first.

File: crates/router/src/connector/paypal/transformers.rs (hunk 18 above)
Comment (swangi-kumari) [reply crates/router/src/connector/paypal/transformers.rs:161]: Initially we had only 1 order, so we are treating whole order as single line items in order

**Review by SanchithHegde — APPROVED** [review]
